_posts/2019-04-02-running-IMA-on-keylime.markdown

```diff
@@ -0,0 +1,449 @@
+
+## Core Root of Trust Measurement (CRTM)
+
+When a platform is rebooted / powered on, all PCRs are set to zero.
+
+On a UEFI system, a series of early boot sequences occur, with one of them being the Core Root of Trust Measurement (CRTM). This is the first measurement written to `PCR: 0`. The CRTM then measures BIOS & boot loader, extending those into `PCR: 0`
+
+## Linux IMA
+
+Early in the OS boot process, Linux IMA (also measured as part of the Kernel in `PCR: 0`) measures (hashes) a policy-based list (`/etc/ima/ima-policy`) of files into `PCR: 10`. The final aggregate hash in `PCR: 10` is then the record of the state of the measured files/directories at time of boot. This record can then be made public (Event Log) and attested using the public key of the TPM.
+
+This attestation mentioned above is achieved using a TPM Quote, where a number of PCRS are hashed, and that hash is signed by a TPM key, accessable only to the TPM.
+
+> Note: Various other PCRs are used for hashing objects such as grubs command line options. 24 PCRs are available in total.
+
+## Wrap up
+
+So with the CRTM based boot process measured and IMA measured files (including runtime access of said files), we have a full attested boot state and runtime integrity environment. Keylime Agent can measure both of these processes and perform a remote attestation using the Keylime Verifier.
+
+For the second section of this blog post, we will go into setting up Keylime to perform a runtime measuring environment, which will sense a non whitelisted script being run by root, which results in the remote Keylime Verifier revoking the node.
+
+## KeyLime IMA
+
+We are going to create two files. A `whitelist.txt` and an `excludes.txt`
+
+## whitelist
+
+The whitelist is a list of what we consider to be golden values (or in fact measurements). We will use the initramfs to create our whitelist using the `create_whitelist.sh` shell script available in th python-keylime repository [here](https://github.com/keylime/python-keylime/blob/master/keylime/create_whitelist.sh) or `python-keylime/keylime/create_whitelist.sh`
+
+We will then remotely send this whitelist to the keylime verifier, which will then proceed to request quotes from the Keylime Agent. Next we will run a non-whitelisted script as root, and see how this results in the verifier revoking the monitored node. For this we use Fedora 29, but other versions /dists can be used ( check with the community on [gitter.im](https://gitter.im/keylime-project/community) for more info.)
+
+Firstly install Keylime using the installer or the ansible playbook (if you need a simulator). Make sure you can communicate with TPM:
+
+```
+# tpm2_getrandom 8
+0x0A 0xBB 0xEB 0xB7 0x82 0x40 0xC1 0x08
+```
+
+On any machine (this could be an unplugged machine for extra security) run the `create_whitelist.sh` script as follows:
+
+```
+# ./create_whitelist.sh ~/whitelist.txt initramfs sha1sum
```

Comment:
opps, I think we chnage this now @jetwhiz

Reply:
Yeah, with https://github.com/keylime/python-keylime/pull/88, create_whitelist.sh will only take up to two arguments (e.g., 'whitelist.txt' and 'sha1sum').